psycopg/connection_type.c

```diff
@@ -126,10 +128,19 @@ psyco_conn_cursor(connectionObject *self, PyObject *args, PyObject *kwargs)
     );
 
     rv = obj;
-    obj = NULL;
+    goto exit;
+
+error:
+    PyErr_Fetch(&error_type, &error_value, &error_traceback);
+    close = PyObject_CallMethod(obj, "close", NULL);
+    if (close)
+        Py_DECREF(close);
+    else
+        PyErr_WriteUnraisable(obj);
```

Comment:
ISMT that if someone passes a factory that doesn't return a `cursor` object, the `IsInstance` test will fail, so we end up in this block, but the `obj.close()` could fail, because that's not a cursor, so the exception raised there would be clobbered.

maybe you should have two objects: `PyObject *obj = null` and `cursorObject *curs null`: once you check `isinstance(obj, cursor)` then you can set `curs = (cursorObject *)obj; obj = NULL;`, use the stronger type for the rest of the function, and use `Py_XDECREF` on both the functions on `exit:` (it's a pattern I prefer instead of tracing carefully the nullness of each var each possible branch of the function).